Add `alias` to category and article and modify default permalinks to include category

Webkadabra opened the pull request.

Webkadabra committed:
Add category alias

Webkadabra committed:
Add `alias`,  `status`, `seo_title`, `seo_keywords` and `seo_description` article attributes

`id` attribute is allowed to be set upon model creation scenario.

Webkadabra committed:
Modify Article::getDetailUrl() to include category alias, if present

Webkadabra committed:
Modify DefaultController::actionDetail to support category aliases in detailed urls

Webkadabra committed:
Update documentation regarding category alias in detailed URLs

Webkadabra committed:
Modify Module's `urlRules` to support category aliases in detailed URLs

Webkadabra committed:
Remove unnecessary information

No need to configure `urlRules` for module in configuration files since new rules are now supported by default.

nadar commented on src/admin/migrations/m180703_183052_add_category_alias.php:
```diff
@@ -0,0 +1,37 @@
+     */
+    public function safeUp()
+    {
+        $this->addColumn('news_cat', 'alias', $this->text());
```
Maybe we could rename to slug? I think the slug field should be varchar `$this->string()` and can not be null `->notNull()`

Webkadabra replied:
I just don't like the sound of a word `slug` so it's never my first choice, but English is my, like, third, language, so if you think slug is a better choice - let's go with it.

nadar replied:
we used alias too, but slug is more analog to yii framework: https://www.yiiframework.com/doc/api/2.0/yii-helpers-baseinflector#slug()-detail

nadar commented on src/admin/migrations/m180703_184214_install_advanced_fields.php:
```diff
@@ -0,0 +1,48 @@
+    public function safeUp()
+    {
+        $this->addColumn('news_article', 'status', "ENUM('trash','published','pending','draft','auto-draft','inherit')");
+        $this->addColumn('news_article', 'alias', $this->text());
```
Same here, maybe we should use `slug` instead and make it varchar and not null.

nadar commented on src/admin/migrations/m180703_184214_install_advanced_fields.php:
```diff
@@ -0,0 +1,48 @@
+    {
+        $this->addColumn('news_article', 'status', "ENUM('trash','published','pending','draft','auto-draft','inherit')");
+        $this->addColumn('news_article', 'alias', $this->text());
+        $this->addColumn('news_article', 'seo_title', $this->text());
```
title should be string not a text

Webkadabra replied:
Will fix!

nadar commented on src/frontend/Module.php:
```diff
@@ -42,6 +42,7 @@
      * @var array
      */
     public $urlRules = [
+        ['pattern' => 'news/<category>/<id:\d+>-<title>', 'route' => 'news/default/detail'],
```
Maybe we should be consistent with the naming along with the other rule:

`news/<category>/<id:\d+>/<title>` with title you mean the alias/slug?

Webkadabra replied:
Yes, but I made it so module does not break if there is no article slug provided - `news` module will continue using generated title for URL

nadar commented on src/models/Article.php:
```diff
@@ -83,8 +88,8 @@ public function rules()
     {
         return [
             [['title', 'text'], 'required'],
```
alias/slug should be required i think.

Webkadabra replied:
I did not want to force usage of category slugs, just add a support. I'm not sure but I think this all should work well for posts that don't have category as well.

nadar replied:
What makes me curious is if people who have to add news will understand this. I think we should have clear version of how to add news, maybe publishing a fork with all your wishes could be a better solution. Its really just a basic news module.

nadar commented on src/models/Article.php:
```diff
@@ -96,7 +101,9 @@ public function rules()
         return [
+            'id' => Module::t('id'),
             'title' => Module::t('article_title'),
+            'alias' => Module::t('article_alias'),
```
the article_alias is missing in the translation file.

Webkadabra replied:
I've added a translation for English, Russian and Ukrainian.

nadar commented on src/models/Article.php:
```diff
@@ -116,9 +123,11 @@ public function attributeLabels()
-            'teaser_text' => ['textarea', 'markdown' => true],
-            'text' => ['textarea', 'markdown' => true],
+            'alias' => 'text',
+            'teaser_text' => ['textarea', 'markdown' => false],
```
Markdown parsing should be enabled in order to support bc.

Webkadabra replied:
Unfortunately, module stops working if markdown is enabled and I've pasted long "complex" html into news body. Not sure what breaks, and it doesn't break all the time. 
What's BC? Breadcrumbs? Maybe I can revert markdown settings back to default but add module option to disable markdown?

nadar replied:
It will break `backward compatibility`, so all the existing news with markdown wont work anymore. so we have to keep markdown enabled.

Webkadabra replied:
Oh, makes sense!

Webkadabra replied:
@nadar oh, makes sense! Sorry, I did not play for many years..

nadar commented on src/models/Article.php:
```diff
@@ -174,8 +190,10 @@ public function ngRestAttributeGroups()
-            [['create', 'update'], ['cat_id', 'title', 'teaser_text', 'text', 'timestamp_create', 'timestamp_display_from', 'is_display_limit', 'timestamp_display_until', 'image_id', 'image_list', 'file_list']],
+            [['list'], ['id', 'cat_id', 'alias', 'title', 'timestamp_create', 'image_id']],
+            // `create` has additional permission to provide `id` for cases when administrator neews to move articles from other blog-modules
+            [['create'], ['id', 'alias', 'cat_id', 'title', 'teaser_text', 'text', 'timestamp_create', 'timestamp_display_from', 'is_display_limit', 'timestamp_display_until', 'image_id', 'image_list', 'file_list']],
```
I think we should stick to all other modules here, entering the id of an PK AI field should be avoided.

Webkadabra replied:
I've added `id` because I want to actually move one of my websites completely to LUYA, and for that I needed a way to provide article ID (for `insert` scenario only). We could have an option on Module level to allow/disable providing of ID when creating an article.

nadar replied:
We can not do this here i think. If you want to import news, why not just create a command very quickly? you could even connect directly to the database. So i don't think its a good approach to just add this element, which makes not that much sense of 99.5% of uses cases.

Webkadabra replied:
@nadar good point!

nadar reviewed (changes requested):
Thanks for helping us improve LUYA! Your pull request will require a "major" release as migration files will break the backwards compatibility. We can do so.

Please take a look at the request change notes.

A changelog entry for 1.1.0 would be required as well.

Webkadabra committed:
Modify field type from `text` to `string`

Webkadabra committed:
Remove `id` from allowed attributes

Webkadabra committed:
Add `enableMarkdown` module option

Some times complex HTML in `Article` may break if `markdown` is enabled.  By default, `markdown` support is enabled, to preserve BC.

Webkadabra committed:
i18n

Webkadabra committed:
i18n

Webkadabra committed:
Rename all `alias` fields to `slug`

`slug` is universally used by Yii: https://www.yiiframework.com/doc/api/2.0/yii-helpers-baseinflector#slug()-detail

Webkadabra commented:
@nadar I reamed `alias` to `slug`

Webkadabra committed:
Fix [error][yii\base\UnknownPropertyException] exception 'yii\base\UnknownPropertyException' with message 'Getting unknown property: luya\cms\admin\Module::enableMarkdown' at uyadev\luya-module-news\src\models\Article.php(129):

Error is thrown when using `news` module at frontend.

nadar commented on src/models/Article.php:
```diff
@@ -115,10 +122,17 @@ public function attributeLabels()
      */
     public function ngRestAttributeTypes()
     {
+        if (Yii::$app->controller->module->id == 'newsadmin') {
```
@Webkadabra i can see the idea behing this, but we should not use fix confiugration inside a model. So maybe we should allow people to override ngrest configurations trough module configuration in general. So we won't have such problems. I was describe such a concept already a long time ago: https://github.com/luyadev/luya-module-admin/blob/master/src/base/Module.php#L63-L95

nadar replied:
So in order to merge the PR, we should just swtich back to the implementation used before.

nadar commented on src/models/Article.php:
```diff
@@ -174,8 +195,9 @@ public function ngRestAttributeGroups()
-            [['list'], ['cat_id', 'title', 'timestamp_create', 'image_id']],
-            [['create', 'update'], ['cat_id', 'title', 'teaser_text', 'text', 'timestamp_create', 'timestamp_display_from', 'is_display_limit', 'timestamp_display_until', 'image_id', 'image_list', 'file_list']],
+            [['list'], ['id', 'cat_id', 'slug', 'title', 'timestamp_create', 'image_id']],
+            [['create'], ['slug', 'cat_id', 'title', 'teaser_text', 'text', 'timestamp_create', 'timestamp_display_from', 'is_display_limit', 'timestamp_display_until', 'image_id', 'image_list', 'file_list']],
```
as create an update are the same now we could write [['create', 'update'], .....] instead of copy the row.

nadar commented:
Thanks for being patient, sometimes its a long process to integrate a feature as there are many aspects we have to take care of.

nadar commented:
I am going to close this PR because the integration seems very complex and we should do such things step by step. I also want to say thank you, for trying to make luya better @Webkadabra we appreciate this!
